ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/RelFieldTrimmer.java

```diff
@@ -0,0 +1,1124 @@
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexPermuteInputsShuttle;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.rex.RexVisitor;
+import org.apache.calcite.sql.SqlExplainFormat;
+import org.apache.calcite.sql.SqlExplainLevel;
+import org.apache.calcite.sql2rel.CorrelationReferenceFinder;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.Bug;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.Pair;
+import org.apache.calcite.util.ReflectUtil;
+import org.apache.calcite.util.ReflectiveVisitor;
+import org.apache.calcite.util.Util;
+import org.apache.calcite.util.mapping.IntPair;
+import org.apache.calcite.util.mapping.Mapping;
+import org.apache.calcite.util.mapping.MappingType;
+import org.apache.calcite.util.mapping.Mappings;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class comes from Calcite almost as-is. The only change concerns
+ * the dispatcher and the builder, so the trimmer is thread-safe and can
+ * be reused across different queries. Definition follows.
+ *
+ * <p>Transformer that walks over a tree of relational expressions, replacing each
+ * {@link RelNode} with a 'slimmed down' relational expression that projects
+ * only the columns required by its consumer.
+ *
+ * <p>Uses multi-methods to fire the right rule for each type of relational
+ * expression. This allows the transformer to be extended without having to
+ * add a new method to RelNode, and without requiring a collection of rule
+ * classes scattered to the four winds.
+ */
+public class RelFieldTrimmer implements ReflectiveVisitor {
+  //~ Static fields/initializers ---------------------------------------------
+
+  private static final Logger LOG = LoggerFactory.getLogger(RelFieldTrimmer.class);
+
+  protected static final ThreadLocal<RelBuilder> REL_BUILDER =
+      new ThreadLocal<>();
+
+  //~ Instance fields --------------------------------------------------------
+
+  protected final HiveReflectUtil.ClassMethodDispatcher<TrimResult, RelNode> trimFieldsDispatcher;
+
+  //~ Constructors -----------------------------------------------------------
+
+  /**
+   * Creates a RelFieldTrimmer.
+   */
+  protected RelFieldTrimmer(boolean useLMFBasedDispatcher) {
+    if (useLMFBasedDispatcher) {
+      this.trimFieldsDispatcher =
+          HiveReflectUtil.createMethodDispatcher(
+              TrimResult.class,
+              this,
+              "trimFields",
+              RelNode.class,
+              ImmutableBitSet.class,
+              Set.class);
+    } else {
+      this.trimFieldsDispatcher =
```

Comment:
Is this still around only for testing/benchmarking? If so a comment would be helpful

Reply:
Added an additional comment here.